hw_4_js/exercises.js

```diff
@@ -0,0 +1,415 @@
+ * [4,3,2,1] => [4,3]
+ */
+
+const get2BiggestValues = (numbers) => {
+  const reducer = ([biggest1, biggest2], n) => {
+    if (biggest1 < n) {
+      biggest2 = biggest1
+      biggest1 = n
+    } else if (biggest2 < n) {
+      biggest2 = n
+    }
+    return [biggest1, biggest2]
+  }
+
+  return numbers.slice(2).reduce(reducer, [numbers[0], numbers[1]])
+}
+
+/**
+ * Exercise 19
+ *
+ * Write a function that returns the number of vowels in a string in English
+ * ( a, e, i, o, u ).
+ *
+ * 'Return the number (count) of vowels in the given string.' => 15
+ */
+
+const getNumberOfVowels = (string) => {
+  return string.match(/[aeiou]/gi).length
+}
+
+/**
+ * Exercise 20
+ *
+ * Write a function that returns an array of two strings where the first element
+ * is the original string with uppercase even letters, and the second
+ * with capital odd.
+ * 'abcdef' => ['AbCdEf', 'aBcDeF']
+ */
+
+const getCapitalizedStrings = (string) => {
+  const mapper = (char, idx) => (idx % 2 != 0 ? char.toLowerCase() : char.toUpperCase())
+  const chars = string.split('')
+
+  const evenCharsCapital = chars.map(mapper).join('')
+  const oddCharsCapital = chars.map((char, idx) => mapper(char, idx + 1)).join('')
+
+  return [evenCharsCapital, oddCharsCapital]
+}
+
+/**
+ * Exercise 21
+ *
+ * Write a function that satisfies the following conditions:
+ *
+ * the function takes a string S, consisting of N letters of the English alphabet in lowercase [a-z]
+ * the function returns a string that does not contain three identical letters in a row
+ * the function removes the minimum number of letters from the string S
+ *
+ * Examples:
+ * S = "eedaaad", the function returns "eedaad". One "a" has been removed.
+ * S = "xxxtxxx", the function returns "xxtxx". The same letters can occur more than three times in a string, but not in a row.
+ * S = "uuuuxaaaaxuuu", the function returns "uuxaaxuu".
+ *
+ * Assumptions:
+ * N is an integer in the range [1..200,000]
+ * S consists only of lowercase letters [a-z]
+ */
+
+const getCorrectString = (string) => {
+  return string.replace(/(.)(?=\1\1)/g, '')
+}
+
+/**
+ * Exercise 22
+ *
+ * Implement a flatten function that takes an array of arbitrary nesting arrays as input
+ * and returns an array of all their elements without nesting.
+ * [1, 2, [3, 4], 5, [[6, 7], 8], 9] => [1, 2, 3, 4, 5, 6, 7, 8, 9]
+ *
+ * or we can just use flat(2) method here instead of
+ */
+
+const getFlattenedArray = (numbers) => {
+  const reducer = (flattenedArray, obj) => flattenedArray.concat(flatten(obj))
+  const flatten = obj => Array.isArray(obj) ? obj.reduce(reducer, []) : obj
+  return numbers.reduce(reducer, [])
+}
+
+/**
+ * Exercise 23
+ *
+ * Implement a function that has an array of numbers as input and an array of not unique values as output.
+ *
+ * [1, 2, 2, 4, 5, 5] => [2, 5]
+ */
+
+const getNotUniqueValues = (numbers) => {
+  return numbers.filter((n, idx) => numbers.indexOf(n) != idx)
```

Comment:
Need to be fixed. Check how it works with `[1, 2, 2, 2, 4, 5, 5, 5] `

Reply:
fixed